File: api/src/routes/plans.ts
import { Router, Request, Response } from 'express';
import { readdir, readFile } from 'fs/promises';
import { join } from 'path';
import { homedir } from 'os';

const router = Router();

function getClaudeDir(): string {
  return join(homedir(), '.claude');
}

// GET /plans - List all plans
router.get('/', async (_req: Request, res: Response) => {
  try {
    const claudeDir = getClaudeDir();
    const plansDir = join(claudeDir, 'plans');

    const files = await readdir(plansDir);
    const planFiles = files.filter(f => f.endsWith('.md'));

    res.json({ data: planFiles });
  } catch {
    res.json({ data: [] });
  }
});

// GET /plans/:planName - Get plan content
router.get('/:planName', async (req: Request, res: Response) => {
  try {
    const { planName } = req.params;
    const claudeDir = getClaudeDir();
    const planPath = join(claudeDir, 'plans', planName);

    // Security: ensure path doesn't escape plans directory
    const normalizedPath = join(claudeDir, 'plans', planName);
    if (!normalizedPath.startsWith(join(claudeDir, 'plans'))) {
      return res.status(400).json({
        code: 'INVALID_PATH',
        message: 'Invalid plan path'
      });
    }

    const content = await readFile(planPath, 'utf-8');

    res.json({
      name: planName,
      content
    });
  } catch {
    res.status(404).json({
      code: 'NOT_FOUND',
      message: 'Plan not found'
    });
  }
});

export { router as plansRouter };

File: api-py/src/routes/plans.py
"""Plans routes for Claude Explorer API."""

from urllib.parse import unquote

from fastapi import APIRouter, HTTPException

from ..utils import get_claude_dir

router = APIRouter(prefix="/plans", tags=["plans"])


@router.get("/")
async def list_plans():
    """List all plans."""
    claude_dir = get_claude_dir()
    plans_dir = claude_dir / "plans"

    if not plans_dir.exists():
        return {"data": []}

    plans = [f.name for f in plans_dir.iterdir() if f.suffix == ".md"]
    return {"data": plans}


@router.get("/{plan_name}")
async def get_plan(plan_name: str):
    """Get a specific plan."""
    plan_name = unquote(plan_name)
    claude_dir = get_claude_dir()
    plan_path = claude_dir / "plans" / plan_name

    # Security check
    if not str(plan_path).startswith(str(claude_dir / "plans")):
        raise HTTPException(status_code=400, detail="Invalid plan name")

    if not plan_path.exists():
        raise HTTPException(status_code=404, detail="Plan not found")

    content = plan_path.read_text()
    return {"name": plan_name, "content": content}
